Adding basic coding guidelines/workflows for the NuGet contributor (#3180)
NuGet/NuGet.Client: docs/coding-guidelines.md

# Coding guidelines

Let's face it. No matter what coding guidelines we choose, we're not going to make everyone happy.
In fact, some people out there might be downright angry at the choices we make.
But the fact of the matter is that there is no "one true bracing style," despite [attempts to name a bracing style as such](http://en.wikipedia.org/wiki/Indent_style#Variant:_1TBS). 

While we would like to embrace everyone's individual style, working together on the same codebase would be utter chaos if we don't enforce some consistency. When it comes to coding guidelines, consistency can be even more important than being "right."

## Copyright header and license notice

> All source code files (mostly `src/**/*.cs` and `test/**/*.cs`) require this exact header (please do not make any changes to it):

Reviewer: ```suggestion
All source code files (mostly `src/**/*.cs` and `test/**/*.cs`) require the _header listed below_ at the beginning of the file, without modifications:
```

```cs
// Copyright (c) .NET Foundation. All rights reserved.
// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
```

> It is ok to skip it on generated files, such as `*.designer.cs`.

Reviewer: It is not mandatory to add the license header to Auto-generated files , such as `*.designer.cs`

Every repo also needs the Apache 2.0 License in a file called LICENSE.txt in the root of the repo. 

> ## C# Coding Style

Reviewer: We need a section for NuGet Error codes

Author: Feel like that goes beyond style, and it's more about you'd write a new feature etc. 
It's good suggestion, but my plan would be for that to get its own section.

1. We only use `var` when it's obvious what the variable type is.
For example the following are correct:

    ```cs
    var fruit = "Apple";
    var fruits = new List<Fruit>();
    string fruit = null; // can't use "var" because the type isn't known (though you could do (string)null, don't!)
    const string expectedName = "name"; // can't use "var" with const
    FruitFlavor flavor = fruit.GetFlavor();
    ```

    The following are incorrect:

    ```cs
    var flavor = fruit.GetFlavor();
    string fruit = "Apple";
    List<Fruit> fruits = new List<Fruit>();
    ```

1. We use language keywords instead of BCL types (e.g. `int, string, float` instead of `Int32, String, Single`, etc) for both type references as well as method calls (e.g. `int.Parse` instead of `Int32.Parse`).
    The following are correct:

    ```cs
    public string TrimString(string s) {

>         return string.IsNullOrEmpty(s)

Reviewer: ```suggestion
    public string TrimString(string s)
    { 
        return string.IsNullOrEmpty(s)
```

Author: Fixed

In constructors where a field or property is initialized coalescing checks are allowed. Otherwise use the regular bracing approach.

```cs

    Path = path ?? throw new ArgumentNullException(nameof(path));
    ....
    if (item == null)
    {
        throw new ArgumentNullException(nameof(item));
    }

```

### Optional parameters

Optional parameters are allowed in certain situations, but discouraged.
Optional parameters are at the root for many versioning issues.
The optional parameter default values are a compile time decision, that's why adding a new optional parameter to an API is a breaking change!
Furthermore, if you have a method with optional parameters, adding an overload with additional optional parameters might cause a compile-time breaking change.
Optional parameters in internal/private methods are acceptable.

### Async method patterns

By default all async methods must have the `Async` suffix. There are some exceptional circumstances where a method name from a previous framework will be grandfathered in.

Passing cancellation tokens can be passed an optional parameter with a value of `default(CancellationToken)`, which is equivalent to `CancellationToken.None`.

> Prefer not having optional parameters.

Reviewer: Why?

Author: See above. 

```
Optional parameters are allowed in certain situations, but discouraged.
Optional parameters are at the root for many versioning issues.
The optional parameter default values are a compile time decision, that's why adding a new optional parameter to an API is a breaking change!
Furthermore, if you have a method with optional parameters, adding an overload with additional optional parameters might cause a compile-time breaking change.
Optional parameters in internal/private methods are acceptable.
```

```cs
Test1
Constructor
FormatString
GetData
```

#### Unit test structure

The contents of every unit test should be split into three distinct stages, optionally separated by these comments:

```cs
// Arrange  
// Act  
// Assert
```

The crucial thing here is that the `Act` stage is exactly one statement. That one statement is nothing more than a call to the one method that you are trying to test. Keeping that one statement as simple as possible is also very important. For example, this is not ideal:

```cs
int result = myObj.CallSomeMethod(GetComplexParam1(), GetComplexParam2(), GetComplexParam3());
```

This style is not recommended because way too many things can go wrong in this one statement. All the `GetComplexParamN()` calls can throw for a variety of reasons unrelated to the test itself. It is thus unclear to someone running into a problem why the failure occurred.

The ideal pattern is to move the complex parameter building into the `Arrange` section:

```cs
// Arrange
P1 p1 = GetComplexParam1();
P2 p2 = GetComplexParam2();
P3 p3 = GetComplexParam3();

// Act
int result = myObj.CallSomeMethod(p1, p2, p3);

// Assert
Assert.AreEqual(1234, result);
```

> 

Reviewer: Is it possible to add a convention that each unit test should test only one part? i.e., don't make a giant unit test that test multiple parts of your code

Author: What do you have in mind? 

Doesn't this cover it?
```
The crucial thing here is that the `Act` stage is exactly one statement. That one statement is nothing more than a call to the one method that you are trying to test. Keeping that one statement as simple as possible is also very important. 
```

Every repo also needs the Apache 2.0 License in a file called LICENSE.txt in the root of the repo. 

## C# Coding Style

The NuGet.Client coding style is similar to the [.NET Framework Coding style](https://github.com/dotnet/runtime/blob/master/docs/coding-guidelines/coding-style.md) with some minor alterations that consider the history of the client repository.

For non code files (xml, etc), our current best guidance is consistency. When editing files, keep new code and changes consistent with the style in the files. For new files, it should conform to the style for that component. If there is a completely new component, anything that is reasonably broadly accepted is fine.

The general rule we follow is "use Visual Studio defaults".

1. We use [Allman style](http://en.wikipedia.org/wiki/Indent_style#Allman_style) braces, where each brace begins on a new line. One exception is that a `using` statement is permitted to be nested within another `using` statement by starting on the following line at the same indentation level, even if the nested `using` contains a controlled block.

1. We use four spaces of indentation (no tabs).

> 1. We use `_camelCase` for internal and private fields and use `readonly` where possible. Prefix internal and private instance fields with `_`. Static fields are all CamelCase regardless of visibility. When used on static fields, `readonly` should come after `static` (e.g. `static readonly` not `readonly static`).  Public fields should be used sparingly and should use PascalCasing with no prefix when used.

Reviewer 2: ```suggestion
1. We use `_camelCase` for internal and private fields and use `readonly` where possible. Prefix internal and private instance fields with `_`. Static fields are all PascalCase regardless of visibility. When used on static fields, `readonly` should come after `static` (e.g. `static readonly` not `readonly static`).  Public fields should be used sparingly and should use PascalCasing with no prefix when used.
```

1. We avoid `this.` unless absolutely necessary.

1. We always specify the visibility, even if it's the default (e.g.
   `private string _foo` not `string _foo`). Visibility should be the first modifier (e.g.
   `public abstract` not `abstract public`).

1. Namespace imports should be specified at the top of the file, *outside* of
   `namespace` declarations, and should be sorted alphabetically, with the exception of `System.*` namespaces, which are to be placed on top of all others.

1. Avoid more than one empty line at any time. For example, do not have two

>    blank lines between members of a type.

Reviewer 2: new line in the middle of the sentence, and no whitespace before next list item.

1. Avoid spurious free spaces.
   For example avoid `if (someVar == 0)...`, where the dots mark the spurious free spaces.

>    Consider enabling "View White Space (Ctrl+R, Ctrl+W)" or "Edit -> Advanced -> View White Space" if using Visual Studio to aid detection.

Reviewer 2: When the markdown is rendered into HTML, the new lines get removed. Given earlier list items have multiple sentences on the same line, it's unclear if the intention here is to have one sentence per line, but in the same HTML paragraph, or if the intention was to have each sentence be rendered as a separate HTML paragraph.

Author: I have it that way to improve readability of the doc.

Author: changed it though for consistency.

1. If a file happens to differ in style from these guidelines (e.g. private members are named `m_member`

>    rather than `_member`), the existing style in that file takes precedence. Changes/refactorings are possible, but depending on the complexity, change frequency of the file, might need to be considered on their own merits in a separate pull request.

Reviewer 2: new line in middle of sentence.

1. We only use `var` when it's obvious what the variable type is.
For example the following are correct:

    ```cs
    var fruit = "Apple";
    var fruits = new List<Fruit>();
    string fruit = null; // can't use "var" because the type isn't known (though you could do (string)null, don't!)
    const string expectedName = "name"; // can't use "var" with const
    FruitFlavor flavor = fruit.GetFlavor();
    ```

    The following are incorrect:

    ```cs
    var flavor = fruit.GetFlavor();
    string fruit = "Apple";
    List<Fruit> fruits = new List<Fruit>();
    ```

1. We use language keywords instead of BCL types (e.g. `int, string, float` instead of `Int32, String, Single`, etc) for both type references as well as method calls (e.g. `int.Parse` instead of `Int32.Parse`).
    The following are correct:

    ```cs
    public string TrimString(string s) {
        return string.IsNullOrEmpty(s)
            ? null
            : s.Trim();

>     }

Reviewer 2: I suggest removing the body of the method, as it doesn't add anything to the example (it doesn't contain any type names), and therefore is identical to the incorrect example, making it harder to see the differences.

Author: It does contain type names: 

string.IsNullOrEmpty(s) vs
String.IsNullOrEmpty(s)

Extension methods are often useful to create chainable method calls, for example, when constructing complex objects, or creating queries.
Internal extension methods are allowed, but bear in mind the previous guideline: ask yourself if an extension method is truly the most appropriate pattern.
The namespace of the extension method class should generally be the namespace that represents the functionality of the extension method, as opposed to the namespace of the target type.
The class name of an extension method container (also known as a "sponsor type") should generally follow the pattern of `<Feature>Extensions`, `<Target><Feature>Extensions`, or `<Feature><Target>Extensions`. For example:

```cs
namespace Food
{
    class Fruit { ... }
}
namespace Fruit.Eating
{
    class FruitExtensions { public static void Eat(this Fruit fruit); }
  OR
    class FruitEatingExtensions { public static void Eat(this Fruit fruit); }
  OR
    class EatingFruitExtensions { public static void Eat(this Fruit fruit); }
}
```

When writing extension methods for an interface the sponsor type name must not start with an `I`.

> ### Doc comments

Reviewer 2: Does this refer to XMLDoc, or docs.microsoft.com?

Author: Fixed it to say Code comments.
